### ISSUE TITLE
HTTP Spicy analyzer - Segmentation Error

### ISSUE THREAD

**Issue by ahgit1** [zeek/spicy-http#1]
I was testing 4.2.0 on centos 7 with the latest spicy analyzers and I had a segmentation error on one of the test pcaps. I disabled the HTTP spicy analyzer and the inbuilt http analyzer worked fine. The pcap I was using was one of the ones which came with the smtp-url-analysis package  (https://github.com/initconf/smtp-url-analysis/blob/master/tests/Traces/smtp-URL-click-many-links.pcap). By trimming the pcap it looked like it was tcp stream 1  causing the error at packet 165 in the overall pcap.

**Comment by timwoj** [comment]
Transferring to the spicy-http repo.

**Comment by bbannier** [comment]
Thank you for the ticket. I tried to reproduce this, but was unable to see a failure on macos (running `zeek -r $PCAP` produced HTTP logs). Since you mention that this was on centos7, it would be great if you could tell us how you set up your environment
- how did you install Zeek
- how did you build and install Spicy (e.g., centos7 only has gcc-8.3 which currently cannot build Spicy; also its version of flex is too old)
- how did you build and install spicy-plugin and spicy-http
- how did you run Zeek

**Comment by ahgit1** [comment]
I have installed the zeek on an offline system so the build was a bit involved - I don't have my notes with me but from what I remember the install was as below, I can get the full details later if you need them. Note: I downloaded the spicy analyzers & plugin on another system and created a zkg bundle  allow installation offline

yum groupinstall "Development Tools"
installed cmake3 and linked to /usr/bin/cmake
installed devtoolset-9
installed rh-python38
installed libmaxminddb
installed libjemalloc-devel
installed spicy rpm (https://github.com/zeek/spicy/releases/download/v1.3.0/spicy_linux_centos_7.rpm)
scl enable devtoolset-9 bash
scl enable rh-python38 bash
PATH=$PATH:/opy/spicy/bin:/usr/local/zeek/bin
tar xvfz zeek-4.2.0.tgz
./configure --enable-jemalloc
make -j
make install
zkg autoconfig
zkg unbundle analyzers.bundle

zeek -NN - shows everything installed correctly at this point
I modified the local.zeek to ignore checksums, and define local networks (I also tried enabling & disabling @packages)  then ran zeek local -r pcap (as root) -  giving the segmentation fault.
I've tried it with lots of other pcaps and it seems to work fine, only giving the error on the one I mentioned

when I disable the spicy http analyzer using 
event zeek_init()
    {
    Spicy::disable_protocol_analyzer(Analyzer::ANALYZER_SPICY_HTTP);
    }
in local.zeek the pcap processes without generating an error

**Comment by bbannier** [comment]
> Note: I downloaded the spicy analyzers & plugin on another system and created a zkg bundle allow installation offline

Was that other system a centos7 system set up in a similar way?

**Comment by ahgit1** [comment]
No - it was actually an ubuntu system - but I took the bundle and unzipped and untarred it, moved the spicy plugin to the top in the manifest and rezipped it, which seemed to allow it to build correctly. I didnt change anything in the analyzer code

**Comment by bbannier** [comment]
We don't have a lot of experience with how Spicy analyzers can be bundled or take any extra steps to make this work, so I am wondering whether we run into an issue here. Spicy analyzers contain compiled C++ code (custom C++ code from analyzer JIT statically linked with Spicy runtime libraries, and also dynamically linked with system libraries), so it don't think we can rule out issues from that.

Could you try again with bundles created on a system that mirrors the target system? That is something we could look into.

**Comment by ahgit1** [comment]
ok - I could take a while doing this as I don't have access to my system at the moment. I was also thinking of manually downloading the analyzers from github and creating my own "bundle" to eliminate the ubuntu step (not sure if that would work, but I'll give it a try). I wasn't aware of the JIT stuff so could well be the issue, Thanks

Added: Just checked the bundle I've been using and none of the code seems to be compiled in there, just .spicy files, so it looks like any compilation will be done when its unbundled (I think)

**Comment by bbannier** [comment]
> Added: Just checked the bundle I've been using and none of the code seems to be compiled in there, just .spicy files, so it looks like any compilation will be done when its unbundled (I think)

You are right, let me try to repro from that. Does `analyzers.bundle` contain just `spicy-plugin` and `spicy-http`?

**Comment by ahgit1** [comment]
No - I put all of the spicy analyzers in there as well 
spicy-plugin
spicy-zip
spicy-tftp
spicy-png
spicy-pe
spicy-ldap
spicy-http
spicy-dns
spicy-dhcp
zeek-spicy-wireguard
zeek-spicy-stun
zeek-spicy-openvpn
zeek-spicy-ipsec
zeek-spicy-fishface
icsnpp-* & zeek-quic - not spicy but were in the bundle

note: spicy-plugin has to be the first item in the manifest.txt

**Comment by bbannier** [comment]
I was able to reproduce this with your description. The bundling seems to be unrelated as this appears to be an ABI-compatibility issues between the C++ libraries on the system and runtime dependencies baked into the Spicy RPM you install.

I see the following triggered when using the prebuild spicy-1.3.0 RPM (originating in the parser code):

```
terminate called after throwing an instance of 'std::logic_error'
  what():  basic_string::_S_construct null not valid
```

If I instead build Spicy on the system inside a `devtoolset-9` environment, the PCAP is analyzed without issues beyond it not conforming to HTTP.

Note that in the meantime we have released spicy-1.4.0 so you might want to use that one instead.
